src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSScoreUtils.java

```diff
@@ -0,0 +1,426 @@
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import scala.Tuple2;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public final class PSScoreUtils {
+
+    public static final String HITS_TAG = "YP";
+    public static final double SCORE_NORMALIZATION_FACTOR = 1e6;
+    private static final Logger logger = LogManager.getLogger(PSScoreUtils.class);
+
+    public static JavaRDD<GATKRead> scoreReads(final JavaSparkContext ctx,
+                                               final JavaRDD<GATKRead> pairedReads,
+                                               final JavaRDD<GATKRead> unpairedReads,
+                                               final SAMFileHeader header,
+                                               final PSScoreArgumentCollection scoreArgs) {
+
+        //Group reads into pairs
+        final JavaRDD<Iterable<GATKRead>> groupedReads = PSScoreUtils.groupReadsIntoPairs(pairedReads,
+                unpairedReads, scoreArgs.readsPerPartition);
+
+        //Load taxonomy database, created by running PathSeqBuildReferenceTaxonomy with this reference
+        final PSTaxonomyDatabase taxDB = PSScoreUtils.readTaxonomyDatabase(scoreArgs.taxonomyDatabasePath);
+        final Broadcast<Map<String, String>> accessionToTaxIdBroadcast = ctx.broadcast(taxDB.accessionToTaxId);
+
+        //Check header against database
+        if (scoreArgs.headerWarningFile != null) {
+            PSScoreUtils.writeHeaderWarnings(scoreArgs.headerWarningFile, header, taxDB, logger);
+        }
+
+        //Determine which alignments are valid hits and return their tax IDs in PSPathogenAlignmentHit
+        //Also adds pathseq tags containing the hit IDs to the reads
+        final JavaRDD<Tuple2<Iterable<GATKRead>, PSPathogenAlignmentHit>> readHits = PSScoreUtils.mapGroupedReadsToTax(groupedReads,
+                scoreArgs.minCoverage, scoreArgs.minIdentity, accessionToTaxIdBroadcast);
+
+        //Collect PSPathogenAlignmentHit objects
+        final Iterable<PSPathogenAlignmentHit> hitInfo = PSScoreUtils.collectTupleSecond(readHits);
+
+        //Get the original reads, now with their pathseq hit tags set
+        final JavaRDD<GATKRead> readsFinal = PSScoreUtils.flattenTupleFirstIterable(readHits);
+
+        //Compute taxonomic scores
+        final Map<String, PSPathogenTaxonScore> taxScores = PSScoreUtils.computeTaxScores(hitInfo, taxDB.tree);
+
+        //Write scores to file
+        PSScoreUtils.writeScoresFile(taxScores, taxDB.tree, scoreArgs.scoresPath);
+
+        return readsFinal;
+    }
+
+    /**
+     * Collects the second elements of all tuples in an RDD
+     */
+    public static <T, C> Iterable<C> collectTupleSecond(final JavaRDD<Tuple2<T, C>> tupleRdd) {
```

Comment:
Also changed flattenTupleSecondIterable to flattenIterableKeys and made package-private